### PR TITLE
vector_typet::size() is now a constant_exprt

### PR DESCRIPTION
**Pull request by kroening**
The codebase assumes basically globally that vector_typet::size() is a
constant_exprt.  This is now enforced by the signature.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by smowton — APPROVED** [review]

File: src/util/std_types.cpp [hunk 1]
```diff
@@ -283,3 +283,19 @@ bool is_constant_or_has_constant_components(
 
   return false;
 }
+
+vector_typet::vector_typet(const typet &_subtype, const constant_exprt &_size)
+  : type_with_subtypet(ID_vector, _subtype)
+{
+  size() = _size;
+}
```
Comment (tautschnig) [inline src/util/std_types.cpp:291]: What's the rationale to move these to the .cpp file?

**Review by tautschnig — APPROVED** [review]
I'm pretty sure this requires further changes in the code base, but of course the stricter type enforcement is good.

File: src/util/std_types.cpp (hunk 1 above)
Comment (kroening) [reply src/util/std_types.cpp:291]: `constant_exprt` isn't complete -- there's a cyclic dependency between std_types.h and std_expr.h.

**Comment by kroening** [comment]
@tautschnig The problem is that there's a period of time in both front-ends where the expression isn't yet turned into a constant. I might have to introduce s.th. like `unfinished_vector_typet` or the like.

**Comment by kroening** [comment]
Added ID_frontend_vector, similar to ID_frontend_pointer, for the type that's yet to be finished.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 41e491392d8f97b8ffd4a4aefa6bed0f1f62b815).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99471857

**Pull request merged**